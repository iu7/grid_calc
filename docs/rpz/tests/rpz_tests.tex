% !TeX document-id = {fb8a2ef5-cdaf-49da-b79d-0a8152e677cd}
% !TeX TS-program = XeLaTeX
\documentclass[a4paper,12pt]{report}

% polyglossia should go first!
\usepackage{polyglossia} % multi-language support
\setmainlanguage{russian}
\setotherlanguage{english}

\usepackage{amsmath} % math symbols, new environments and stuff
\usepackage{unicode-math} % for changing math font and unicode symbols
\usepackage[style=english]{csquotes} % fancy quoting
\usepackage{microtype} % for better font rendering
\usepackage{hyperref} % for refs and URLs
\usepackage{graphicx} % for images (and title page)
\usepackage{geometry} % for margins in title page
\usepackage{tabu} % for tabulars (and title page)
\usepackage[section]{placeins} % for float barriers
\usepackage{titlesec} % for section break hooks
\usepackage{listings} % for listings 
\usepackage{upquote} % for good-looking quotes in source code (used for custom languages)
\usepackage{xcolor} % colors!
\usepackage{enumitem} % for unboxed description labels (long ones)
\usepackage{caption}
\usepackage{multirow}
\usepackage{varwidth}
\usepackage{longtable}

\defaultfontfeatures{Mapping=tex-text} % for converting "--" and "---"
\setmainfont{CMU Serif}
\setsansfont{CMU Sans Serif}
\setmonofont{CMU Typewriter Text}
\setmathfont{XITS Math}
\MakeOuterQuote{"} % enable auto-quotation

% new page and barrier after section, also phantom section after clearpage for
% hyperref to get right page.
% clearpage also outputs all active floats:
\newcommand{\sectionbreak}{\phantomsection}
\newcommand{\subsectionbreak}{\FloatBarrier}
\renewcommand{\thesection}{\arabic{section}} % no chapters
\numberwithin{equation}{section}
%\usetikzlibrary{shapes,arrows,trees}

\newcommand{\itemtt}[1][]{\item[\texttt{#1}:]} % tt-ed items (for protocol descriptions)

\definecolor{bluekeywords}{rgb}{0.13,0.13,1}
\definecolor{greencomments}{rgb}{0,0.5,0}
\definecolor{turqusnumbers}{rgb}{0.17,0.57,0.69}
\definecolor{redstrings}{rgb}{0.5,0,0}
\setmonofont{Consolas} %to be used with XeLaTeX or LuaLaTeX
\definecolor{bluekeywords}{rgb}{0,0,1}
\definecolor{greencomments}{rgb}{0,0.5,0}
\definecolor{redstrings}{rgb}{0.64,0.08,0.08}
\definecolor{xmlcomments}{rgb}{0.5,0.5,0.5}
\definecolor{types}{rgb}{0.17,0.57,0.68}

\lstloadlanguages{bash, python, Java}

\lstset{
  frame=none,
  xleftmargin=2pt,
  stepnumber=1,
  numbers=left,
  numbersep=5pt,
  numberstyle=\ttfamily\tiny\color[gray]{0.3},
  belowcaptionskip=\bigskipamount,
  captionpos=b,
  escapeinside={*'}{'*},
  language=python,
  tabsize=2,
  emphstyle={\bf},
  commentstyle=\it,
  stringstyle=\mdseries\rmfamily,
  showspaces=false,
  keywordstyle=\bfseries\rmfamily,
  columns=flexible,
  basicstyle=\small\sffamily,
  showstringspaces=false,
  morecomment=[l]\%,
  breaklines=true,
  showlines=true
}
\renewcommand\lstlistingname{Листинг}

\date{\today}

\makeatletter
\let\thetitle\@title
\let\theauthor\@author
\let\thedate\@date
\makeatother

\makeatletter
\AtBeginDocument{%
    \expandafter\renewcommand\expandafter\subsection\expandafter{%
        \expandafter\@fb@secFB\subsection
    }%
}
\makeatother

\begin{document}

\section{Тестирование}
\subsection{Модульное тестирование}
Статический класс DataMethods:


Представляет собой интерфейс бекенда данных.


Производится тестирование класса DataMethods с использованием методики тестирования -- разбиение на уровне класса на категории по функциональности. Категория объединяет в себе методы класса, выполняющие близкую по смыслу функциональность.


Методы класса можно разбить на 3 категории по функциональности:
\begin{itemize}
  \item методы получения данных;
  \item методы изменения данных;
  \item методы удаления данных. 
\end{itemize}

\begin{table}[h!]
\caption{Методы класса DataMethodsFilter}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название метода & Примечания \\
\hline
\multirow{3}{*}{get\_item} & param: table [ str ] \\
                           & param: value\_json[ dict ] \\
                           & Возвращает все элементы таблицы с именем table, удовлетворяющие фильтру value\_json \\
\hline
\multirow{3}{*}{put\_item} & param: table [ str ] \\
                           & param: value\_json [ dict ] \\
                           & Изменяет все элементы таблицы с именем table, удовлетворяющие фильтру value\_json \\
\hline
\multirow{3}{*}{delete\_item} & param: table [ str ] \\
                           & param: value\_json [ dict ] \\
                           & Удаляет все элементы таблицы с именем table, удовлетворяющие фильтру value\_json \\
\hline
\end{tabu}
\end{table}

%%%
\clearpage
%%%

\begin{table}[h]
\caption{Категория 1 -- Тестирование метода получения данных}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название теста & TestGetItemEmpty \\ \hline
Тестируемый метод & get\_item \\ \hline
Описание теста & Проверка получения данных из БД, пустой таблицы trait \\ \hline
Ожидаемый результат & Cловарь с пустым списком в ключе ``result'' \\ \hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%

\begin{table}[h]
\caption{Категория 1 -- Тестирование метода получения данных}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название теста & TestGetItemGeneral \\ \hline
Тестируемый метод & get\_item \\ \hline
\multirow{4}{*}{Описание теста} & Проверка изменения данных в БД, в таблице trait, предварительно наполненной записями \\
                                & name = ``test\_name'', version = ``1.0'' \\
                                & name = ``test\_name'', version = ``2.0'', \\
                                & по фильтру \{``name'':``test\_name''\} \\
\hline
\multirow{3}{*}{Ожидаемый результат} & Словарь \{``result'':[\{``name'':``test\_name'', \\
                                & ``version'':``1.0''\}, \{``name'':``test\_name'',\} \\
                                & ``version'':``2.0''\}]\} \\
\hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%
\clearpage
%%%

\begin{table}[h]
\caption{Категория 2 -- Тестирование метода изменения данных}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название теста & TestPutItem \\ \hline
Тестируемый метод & put\_item \\ \hline
\multirow{4}{*}{Описание теста} & Проверка изменения данных в БД, в таблице trait, предварительно наполненной записями \\
                                & name = ``test\_name'', version = ``1.0'' \\
                                & name = ``test\_name'', version = ``2.0'', \\
                                & по фильтру \{``name'':``test\_name'',``changes'': \{``version'':``3.0''\}\} \\
\hline
Ожидаемый результат & Словарь \{``count'': 2\}, отражающий наличие двух произведенных изменений в БД \\ \hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%

\begin{table}[h]
\caption{Категория 3 -- Тестирование метода удаления данных}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название теста & TestDeleteItem \\ \hline
Тестируемый метод & delete\_item \\ \hline
\multirow{4}{*}{Описание теста} & Проверка удаления данных в БД, в таблице trait, предварительно наполненной записями \\
                                & name = ``test\_name'', version = ``1.0'' \\
                                & name = ``test\_name'', version = ``2.0'', \\
                                & по фильтру \{``name'':``test\_name''\} \\
\hline
Ожидаемый результат & Словарь \{``count'': 2\}, отражающий наличие двух произведенных удалений в БД \\ \hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}


Вывод по результатам тестирования:\\
Все тесты пройдены успешно, класс готов к использованию.

\clearpage
\subsection{Интеграционное тестирование}
Производится интеграционное тестирование подсистем работы с файлами, данными и подсистемы мониторинга. Целью тестирования явлеяется проверка
корректности регистрации подсистем на маяке, играющем ключевую роль в работе распределленной системы.

Интерфейс подсистемы мониторинга(категория -- beacon, сокращенно -- маяк):
\begin{itemize}
  \item GET /services/<service\_group>
  \item PUT /services/<service\_group>/<service\_host>:<service\_port>
  \item POST /services
\end{itemize}

Методы подсистемы работы с файлами(категория -- filesystem):
\begin{itemize}
  \item beacon\_setter
  \item beacon\_getter
\end{itemize}

Методы подсистемы работы с данными(категория -- database):
\begin{itemize}
  \item beacon\_setter
  \item beacon\_getter
\end{itemize}

\begin{table}[h]
\caption{Тестирование регистрации бекендов на не запущенном маяке}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название файла & TestNoBeaconGetter.sh \\ \hline
Взаимодействующие подсистемы & database, filesystem, beacon \\ \hline
Описание теста & Подсистемы filesystem и database выполняют метод beacon\_setter при запуске \\ \hline
Начальные условия & beacon не запущен \\
Ожидаемый результат & подсистемы не могут подключиться к beacon и сообщают об этом пользователю; повторный поиск происходит регулярно, с интервалом в 10 секунд \\ \hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%
\clearpage
%%%

\begin{table}[H]
\caption{Тестирование регистрации бекендов на запущенном маяке}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название файла & TestBeaconPoster.sh \\ \hline
Взаимодействующие подсистемы & database, filesystem, beacon \\ \hline
Описание теста & Подсистемы filesystem и database выполняют метод beacon\_setter при запуске \\ \hline
Начальные условия & beacon запущен \\
\multirow{4}{*}{Ожидаемый результат} & Подсистемы подключаются к маяку и выполняют POST-запрос на адрес /services/<категория бекенда>, 
                                       передавая через JSON свою адрес \\
                                     & Журнал маяка: \\
                                     & POST /services/database HTTP/1.1 200 \\
                                     & POST /services/filesystem HTTP/1.1 200 \\
\hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%

\begin{table}[H]
\caption{Тестирование регистрации бекендов на запущенном маяке}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название файла & TestBeaconPutter.sh \\ \hline
Взаимодействующие подсистемы & database, filesystem, beacon \\ \hline
Описание теста & Подсистемы filesystem и database выполняют метод beacon\_setter повторно \\ \hline
Начальные условия & beacon запущен, подсистемы уже выполнили первичный POST запрос \\
\multirow{5}{*}{Ожидаемый результат} & Подсистемы подключаются к маяку и выполняют PUT-запрос на адрес \\
                                     & /services/<категория бекенда>/ \\
                                     &<адрес бекенда>:<порт>. \\
                                     & Пример журнала маяка: \\
                                     & PUT /services/database/localhost:5000 HTTP/1.1 200 \\
                                     & PUT /services/filesystem/localhost:5001 HTTP/1.1 200 \\
\hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}

%%%
\clearpage
%%%

\begin{table}[h]
\caption{Тестирование регистрации бекендов на запущенном маяке}
\begin{tabu} to \textwidth {|c|X|}
\hline
Название файла & TestBeaconGetter.sh \\ \hline
Взаимодействующие подсистемы & filesystem, beacon \\ \hline
Описание теста & Подсистема filesystem запрашивает адреса всех подсистем database у маяка \\ \hline
Начальные условия & beacon запущен, подсистемы уже выполнили первичный POST запрос \\
\multirow{2}{*}{Ожидаемый результат} & filesystem подключаются к маяку и выполняет GET-запрос(метод beacon\_getter)
                                       на адрес /services/database/. \\
                                      & Маяк отвечает все известные адреса бекендов, категории database
                                       в JSON-формате. Пример ответа -- [\{``localhost'': 5000\}] \\
\hline
Степень важности & Фатальная \\ \hline
Результат теста & Тест пройден \\ \hline
\end{tabu}
\end{table}


Вывод:


Интеграционное тестирование выявило фатальную ошибку в реализации подсистемы beacon, связанную с неверно документированным
поведением метода доступа к элементу по ссылке в словарях многократной вложенности в языке Python.


Ошибка была устранена с использованием альтернативной библиотченой реализации этого метода.
\end{document}