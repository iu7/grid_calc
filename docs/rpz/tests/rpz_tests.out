\BOOKMARK [1][-]{section*.1}{Тестирование}{}% 1
\BOOKMARK [2][-]{subsection.0.1.1}{Модульное тестирование}{section*.1}% 2
\BOOKMARK [2][-]{subsection.0.1.2}{Интеграционное тестирование}{section*.1}% 3
